### PR TITLE
Fix #42: Serious gltf bugs

### LINKED ISSUE #42: Serious glTF importer bugs
@Squareys @NussknackerXXL

I'm quite disappointed by state of the glTF importer code, so the tone of the following might not be all :white_flower: and :sun_with_face:. Sorry in advance.

As reported by https://twitter.com/steeve/status/997138966754230272, the glTF importer plugin at its current state **doesn't seem to work at all** for *any* model in https://github.com/KhronosGroup/glTF-Sample-Models. I wonder what it was tested on? 

I managed to fix it to display at least the mesh data correctly for Box and and AnimatedBox (patch below, not commited anywhere as it *needs* corresponding tests), but it still fails *really bad* on all other models. For the Cesium Milk Truck only the wheels are displayed and with weird texture coordinates, for example. Here's a patch that fixes at least something:

```diff
diff --git a/src/MagnumPlugins/TinyGltfImporter/TinyGltfImporter.cpp b/src/MagnumPlugins/TinyGltfImporter/TinyGltfImporter.cpp
index 690ddc0..cf8f7b2 100644
--- a/src/MagnumPlugins/TinyGltfImporter/TinyGltfImporter.cpp
+++ b/src/MagnumPlugins/TinyGltfImporter/TinyGltfImporter.cpp
@@ -317,9 +317,9 @@ Containers::Optional<MeshData3D> TinyGltfImporter::doMesh3D(const UnsignedInt id
                 return Containers::NullOpt;
             }
 
-            const size_t numPositions = bufferView.byteLength/sizeof(Vector3);
+            const size_t numPositions = accessor.count;
             positions.reserve(numPositions);
-            std::copy_n(reinterpret_cast<const Vector3*>(buffer.data.data() + bufferView.byteOffset), numPositions, std::back_inserter(positions));
+            std::copy_n(reinterpret_cast<const Vector3*>(buffer.data.data() + bufferView.byteOffset + accessor.byteOffset), numPositions, std::back_inserter(positions));
 
         } else if(attribute.first == "NORMAL") {
             if(accessor.type != TINYGLTF_TYPE_VEC3) {
@@ -327,9 +327,9 @@ Containers::Optional<MeshData3D> TinyGltfImporter::doMesh3D(const UnsignedInt id
                 return Containers::NullOpt;
             }
 
-            const size_t numNormals = bufferView.byteLength/sizeof(Vector3);
+            const size_t numNormals = accessor.count;
             normals.reserve(numNormals);
-            std::copy_n(reinterpret_cast<const Vector3*>(buffer.data.data() + bufferView.byteOffset), numNormals, std::back_inserter(normals));
+            std::copy_n(reinterpret_cast<const Vector3*>(buffer.data.data() + bufferView.byteOffset + accessor.byteOffset), numNormals, std::back_inserter(normals));
 
         /* Texture coordinate attribute ends with _0, _1 ... */
         } else if(Utility::String::beginsWith(attribute.first, "TEXCOORD")) {
@@ -341,9 +341,9 @@ Containers::Optional<MeshData3D> TinyGltfImporter::doMesh3D(const UnsignedInt id
             textureLayers.emplace_back();
             std::vector<Vector2>& textureCoordinates = textureLayers.back();
 
-            const size_t numTextureCoordinates = bufferView.byteLength/sizeof(Vector2);
+            const size_t numTextureCoordinates = accessor.count;
             textureCoordinates.reserve(numTextureCoordinates);
-            std::copy_n(reinterpret_cast<const Vector2*>(buffer.data.data() + bufferView.byteOffset), numTextureCoordinates, std::back_inserter(textureCoordinates));
+            std::copy_n(reinterpret_cast<const Vector2*>(buffer.data.data() + bufferView.byteOffset + accessor.byteOffset), numTextureCoordinates, std::back_inserter(textureCoordinates));
 
         /* Color attribute ends with _0, _1 ... */
         } else if(Utility::String::beginsWith(attribute.first, "COLOR")) {
@@ -383,13 +383,13 @@ Containers::Optional<MeshData3D> TinyGltfImporter::doMesh3D(const UnsignedInt id
     }
 
     std::vector<UnsignedInt> indices;
-    const UnsignedByte* start = idxBuffer.data.data() + idxBufferView.byteOffset;
+    const UnsignedByte* start = idxBuffer.data.data() + idxBufferView.byteOffset + idxAccessor.byteOffset;
     if(idxAccessor.componentType == TINYGLTF_COMPONENT_TYPE_UNSIGNED_BYTE) {
-        std::copy_n(start, idxBufferView.byteLength, std::back_inserter(indices));
+        std::copy_n(start, idxAccessor.count, std::back_inserter(indices));
     } else if(idxAccessor.componentType == TINYGLTF_COMPONENT_TYPE_UNSIGNED_SHORT) {
-        std::copy_n(reinterpret_cast<const UnsignedShort*>(start), idxBufferView.byteLength/sizeof(UnsignedShort), std::back_inserter(indices));
+        std::copy_n(reinterpret_cast<const UnsignedShort*>(start), idxAccessor.count, std::back_inserter(indices));
     } else if(idxAccessor.componentType == TINYGLTF_COMPONENT_TYPE_UNSIGNED_INT) {
-        std::copy_n(reinterpret_cast<const UnsignedInt*>(start), idxBufferView.byteLength/sizeof(UnsignedInt), std::back_inserter(indices));
+        std::copy_n(reinterpret_cast<const UnsignedInt*>(start), idxAccessor.count, std::back_inserter(indices));
     } else CORRADE_ASSERT_UNREACHABLE(); /* LCOV_EXCL_LINE */
 
     return MeshData3D(meshPrimitive, std::move(indices), {std::move(positions)}, {std::move(normals)}, textureLayers, colorLayers, &mesh);

```

So, what's needed:

- [x] look at [this tutorial](https://github.com/KhronosGroup/glTF-Tutorials/blob/master/gltfTutorial/gltfTutorial_005_BuffersBufferViewsAccessors.md) and understand how buffers and views and accessors work. The code had it *totally* wrong and the test data did not cover *any of that* at all.
- [x] As far as I know, the official glTF repositories provide test data for all the corner cases. Use them, please.
- [x] Only the 8-bit indices are tested, [never the 16-bit or 32-bit indices](https://codecov.io/gh/mosra/magnum-plugins/src/master/src/MagnumPlugins/TinyGltfImporter/TinyGltfImporter.cpp) (and they have the same problem re views and accessors). These need to be fixed *and tested* too, the patch does that a bit, but probably not fully.
- [x] Assimp is able to import at least some material properties (color and roughness from PBR materials), so maybe a similar thing could be done here (until we have PBR import). Otherwise we get more reports about everything being plain gray.
- [x] In #41 [I asked for](https://github.com/mosra/magnum-plugins/pull/41#issuecomment-387343478) *testing and fixing* proper handling of file opening failure. It was not done. Opening a nonexistent file still blows up somewhere later in the code. (I admit I should have triple checked before merging, sorry.)

**This is quite important**, as the latest release was marketed as having glTF support. This is far from that. I hope you have time for this, otherwise I need to drop whatever I am doing now and take over this plugin, as a lot of things will rely on it in the near future.

### PR DESCRIPTION
**Pull request by Squareys**
Hi @mosra !

Here's a WIP PR for #42!

Cheers, Jonathan.

**TODOs**
 - [x] bufferView.stride (now failing on non-trivial as suggested)
 - [x] sensible material defaults (and interpretation of pbr values)
 - [x] fix flipped textures
 - [x] solve python scripts/buffer generation
 - [x] test load more of the khronos provided example models

### CODE REVIEW

File: src/MagnumPlugins/TinyGltfImporter/TinyGltfImporter.cpp [hunk 1]
```diff
@@ -222,43 +224,50 @@ std::unique_ptr<ObjectData3D> TinyGltfImporter::doObject3D(UnsignedInt id) {
     CORRADE_INTERNAL_ASSERT(node.rotation.size() == 0 || node.rotation.size() == 4);
     CORRADE_INTERNAL_ASSERT(node.translation.size() == 0 || node.translation.size() == 3);
     CORRADE_INTERNAL_ASSERT(node.scale.size() == 0 || node.scale.size() == 3);
+    /* Ensure we have either a matrix or L-R-S */
```
Comment (mosra) [inline src/MagnumPlugins/TinyGltfImporter/TinyGltfImporter.cpp]: What does L stand for? (The glTF tutorial names it as TRS.)

File: src/external/tinygltf/tiny_gltf.h [hunk 2]
```diff
@@ -4,7 +4,7 @@
 //
 // The MIT License (MIT)
 //
-// Copyright (c) 2015 - 2017 Syoyo Fujita, Aurélien Chatelain and many
+// Copyright (c) 2015 - 2018 Syoyo Fujita, Aurélien Chatelain and many
```
Comment (mosra) [inline src/external/tinygltf/tiny_gltf.h]: Can you amend the commit message that updates this file and say which upstream SHA-1 is this? Helps a lot when updating later.

Thanks.

File: src/MagnumPlugins/TinyGltfImporter/TinyGltfImporter.cpp [hunk 3]
```diff
@@ -288,6 +297,9 @@ Containers::Optional<MeshData3D> TinyGltfImporter::doMesh3D(const UnsignedInt id
         meshPrimitive = MeshPrimitive::Lines;
     } else if(primitive.mode == TINYGLTF_MODE_LINE_LOOP) {
         meshPrimitive = MeshPrimitive::LineLoop;
+    } else if(primitive.mode == 3) {
+        /* For some reason tiny_gltf doesn't have a define for this */
```
Comment (mosra) [inline src/MagnumPlugins/TinyGltfImporter/TinyGltfImporter.cpp]: :see_no_evil: :laughing:

File: src/MagnumPlugins/TinyGltfImporter/TinyGltfImporter.cpp [hunk 4]
```diff
@@ -317,19 +329,19 @@ Containers::Optional<MeshData3D> TinyGltfImporter::doMesh3D(const UnsignedInt id
                 return Containers::NullOpt;
             }
 
-            const size_t numPositions = bufferView.byteLength/sizeof(Vector3);
+            const size_t numPositions = accessor.count;
```
Comment (mosra) [inline src/MagnumPlugins/TinyGltfImporter/TinyGltfImporter.cpp]: I think this variable is not really needed anymore. (Yes, sorry, this originates from my patch, which was just hastily put together to make things working).

File: src/MagnumPlugins/TinyGltfImporter/TinyGltfImporter.cpp [hunk 5]
```diff
@@ -414,53 +426,62 @@ std::unique_ptr<AbstractMaterialData> TinyGltfImporter::doMaterial(const Unsigne
...
+        auto shininessFactorValue = cmnBlinnPhongExt.Get("shininessFactor");
+        if(shininessFactorValue.Type() != tinygltf::NULL_TYPE) {
+            shininess = float(shininessFactorValue.Get<double>());
+        }
+    } else {
+        auto dt = material.values.find("baseColorTexture");
+        if(dt != material.values.end()) {
+            diffuseTexture = dt->second.TextureIndex();
+            flags |= PhongMaterialData::Flag::DiffuseTexture;
+        }
```
Comment (mosra) [inline src/MagnumPlugins/TinyGltfImporter/TinyGltfImporter.cpp]: Similarly to `baseColorTexture`, there's the `baseColorFactor`. As I mentioned in #42, the glTF sample model repository uses it on even the most basic models. Assimp imports it as diffuse, so we could do the same here (unless `diffuseFactor` is set).

File: src/MagnumPlugins/TinyGltfImporter/TinyGltfImporter.cpp [hunk 6]
```diff
@@ -383,13 +395,13 @@ Containers::Optional<MeshData3D> TinyGltfImporter::doMesh3D(const UnsignedInt id
...
     if(idxAccessor.componentType == TINYGLTF_COMPONENT_TYPE_UNSIGNED_BYTE) {
-        std::copy_n(start, idxBufferView.byteLength, std::back_inserter(indices));
+        std::copy_n(start, idxAccessor.count, std::back_inserter(indices));
     } else if(idxAccessor.componentType == TINYGLTF_COMPONENT_TYPE_UNSIGNED_SHORT) {
-        std::copy_n(reinterpret_cast<const UnsignedShort*>(start), idxBufferView.byteLength/sizeof(UnsignedShort), std::back_inserter(indices));
+        std::copy_n(reinterpret_cast<const UnsignedShort*>(start), idxAccessor.count, std::back_inserter(indices));
     } else if(idxAccessor.componentType == TINYGLTF_COMPONENT_TYPE_UNSIGNED_INT) {
-        std::copy_n(reinterpret_cast<const UnsignedInt*>(start), idxBufferView.byteLength/sizeof(UnsignedInt), std::back_inserter(indices));
+        std::copy_n(reinterpret_cast<const UnsignedInt*>(start), idxAccessor.count, std::back_inserter(indices));
     } else CORRADE_ASSERT_UNREACHABLE(); /* LCOV_EXCL_LINE */
```
Comment (mosra) [inline src/MagnumPlugins/TinyGltfImporter/TinyGltfImporter.cpp]: Unless I missed something, this is still not correctly handling [interleaved data](https://github.com/KhronosGroup/glTF-Tutorials/blob/master/gltfTutorial/gltfTutorial_005_BuffersBufferViewsAccessors.md#data-interleaving). Failing with a clear error in that case would be the best, but can you mention that in the limitations in the docs, at least? I have no idea if tinygltf supports that, even.

File: src/MagnumPlugins/TinyGltfImporter/TinyGltfImporter.cpp [hunk 7]
```diff
@@ -473,6 +494,12 @@ UnsignedInt TinyGltfImporter::doTextureCount() const {
 
 Containers::Optional<TextureData> TinyGltfImporter::doTexture(const UnsignedInt id) {
     const tinygltf::Texture& tex = _d->model.textures[id];
+
+    if(tex.sampler < 0) {
+        /* The specification instructs to use "auto sampling", i.e. it is left to the implementor to decide on the default values... */
+        return TextureData{TextureData::Type::Texture2D, SamplerFilter::Linear, SamplerFilter::Linear,
+            Sampler::Mipmap::Linear, {Sampler::Wrapping::Repeat, Sampler::Wrapping::Repeat, Sampler::Wrapping::Repeat}, UnsignedInt(tex.source), &tex};
+    }
```
Comment (mosra) [inline src/MagnumPlugins/TinyGltfImporter/TinyGltfImporter.cpp]: Would be great if you could mention the defaults in case no sampler definition is present in the docs.

File: src/MagnumPlugins/TinyGltfImporter/TinyGltfImporter.cpp [hunk 8]
```diff
@@ -414,53 +426,62 @@ std::unique_ptr<AbstractMaterialData> TinyGltfImporter::doMaterial(const Unsigne
     /* Textures */
     PhongMaterialData::Flags flags;
     UnsignedInt diffuseTexture{}, specularTexture{};
+    Vector4 diffuseColor{1.0f};
+    Vector3 specularColor{0.0f};
+    Float shininess{1.0f};
```
Comment (mosra) [inline src/MagnumPlugins/TinyGltfImporter/TinyGltfImporter.cpp]: Mentioning these defaults in the docs would be great :)

**Comment by mosra** [comment]
Oh, one more thing: are you planning to handle the texture coordinate flipping here as well? I am not exactly sure what's happening, though.

File: src/MagnumPlugins/TinyGltfImporter/Test/mesh-primitives.bin.in [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+unsigned int: 0 1
+# Line Strip
+float: 2.1 2.2 2.4 2.5 2.6 2.7 2.8 2.9 3.0
+unsigned int: 2 1 0
+# Triangle Strip
+float: 3.1 3.2 3.4 3.5 3.6 3.7 3.8 3.9 4.0 4.1 4.2 4.3
+unsigned int: 2 1 0 3
+# Triangle Fan
+float: 5.1 5.2 5.3 6.1 6.2 6.3 7.1 7.2 7.3 8.1 8.2 8.3
+unsigned int: 2 1 3 0
```
[the remark this answers is not in the document]
Comment (Squareys) [reply src/MagnumPlugins/TinyGltfImporter/Test/mesh-primitives.bin.in]: Back when I read that message I thought you meant having that comment like in the code snippet which would then be parsed and then insert its result into the appropriate place in the gltf file :D That seemed a little too complicated.
Didn't realize you meant having a script that would just create the bin in a hardcoded way... kinda makes sense, though.

File: src/MagnumPlugins/TinyGltfImporter/TinyGltfImporter.cpp (hunk 1 above)
Comment (Squareys) [reply src/MagnumPlugins/TinyGltfImporter/TinyGltfImporter.cpp]: Location. (Like LocRotScale in blender)

Will change it to match the spec

File: src/MagnumPlugins/TinyGltfImporter/TinyGltfImporter.cpp (hunk 6 above)
Comment (Squareys) [reply src/MagnumPlugins/TinyGltfImporter/TinyGltfImporter.cpp]: It does, there is `accessor.stride`...

Comment (Squareys) [reply src/MagnumPlugins/TinyGltfImporter/TinyGltfImporter.cpp]: Pardon me, it's bufferView.stride (just in case you need it sometime)

File: src/MagnumPlugins/TinyGltfImporter/TinyGltfImporter.cpp [hunk 10]
```diff
@@ -320,6 +320,9 @@ Containers::Optional<MeshData3D> TinyGltfImporter::doMesh3D(const UnsignedInt id
         const tinygltf::BufferView& bufferView = _d->model.bufferViews[accessor.bufferView];
         const tinygltf::Buffer& buffer = _d->model.buffers[bufferView.buffer];
 
+        const int stride = bufferView.byteStride;
+        CORRADE_ASSERT(stride == 0, "Trade::TinyGltfImporter::mesh3D(): non-zero strides for buffer views are unsupported.", {});
```
Comment (mosra) [inline src/MagnumPlugins/TinyGltfImporter/TinyGltfImporter.cpp]: ... can this be a non-fatal failure? Like, returning `NullOpt` instead of blowing up :)

Comment (Squareys) [reply src/MagnumPlugins/TinyGltfImporter/TinyGltfImporter.cpp]: Yes, already did that.

**Comment by Squareys** [comment]
@mosra Alright, did everything (I tripple checked this time, but you may want to do so, too 😜 ). Ready for review 👍

**Comment by mosra** [comment]
Finally merged, thank you a lot :)

**Comment by steeve** [comment]
congrats guys ! and thank you !
